src/FakeItEasy.Analyzer/NonVirtualSetupAnalyzer.cs

```diff
@@ -0,0 +1,108 @@
+﻿namespace FakeItEasy.Analyzer
+{
+    using System;
+    using System.Collections.Immutable;
+    using System.Linq;
+    using Microsoft.CodeAnalysis;
+    using Microsoft.CodeAnalysis.CSharp;
+    using Microsoft.CodeAnalysis.CSharp.Syntax;
+    using Microsoft.CodeAnalysis.Diagnostics;
+
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    class NonVirtualSetupAnalyzer : DiagnosticAnalyzer
+    {
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(DiagnosticDefinitions.NonVirtualSetup);
+
+        private static readonly ImmutableDictionary<string, DiagnosticDescriptor> DiagnosticsMap = CreateDiagnosticsMap();
+
+        public override void Initialize(AnalysisContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            context.RegisterSyntaxNodeAction(AnalyzeMethod,
+                SyntaxKind.InvocationExpression);
+
+            context.RegisterSyntaxNodeAction(AnalyzeProperty,
+                SyntaxKind.SimpleMemberAccessExpression);
+        }
+
+        public static void AnalyzeMethod(SyntaxNodeAnalysisContext context)
+        {
+            //Method so exclude non-methods and interfaces
+
+            AnalyzeCall<InvocationExpressionSyntax>(context, x => NotMethod(x) || Interface(x));
+        }
+
+        public static void AnalyzeProperty(SyntaxNodeAnalysisContext context)
+        {
+            //Propery so exclude methods
+
+            AnalyzeCall<MemberAccessExpressionSyntax>(context, x => Method(x));
+        }
+
+        private static void AnalyzeCall<T>(SyntaxNodeAnalysisContext context, Func<SymbolInfo, bool> exclusions) where T : ExpressionSyntax

```

Comment:
Rather than excluding items to examine, could we not just include items?
That way, rather than saying the property analyzer excludes methods, and the method analyzer excludes non-methods, we could just say that the property analyzer applies to properties and the method analyzer applies to methods?

The `Interface` portion of the exclusion could be placed right into this method, where it would be run whatever we're analyzing (which would eliminate the problem I mention with `AnalyzeProperty` on line 43.